File: schemas/application_schema.py
from marshmallow import (
    Schema,
    fields,
    validates,
    ValidationError
)

from datetime import date


class ApplicationSchema(Schema):

    company = fields.Str(required=True)

    role = fields.Str(required=True)

    applied_date = fields.Date(required=False)

    @validates("company")
    def validate_company(self, value, **kwargs):

        if not value.strip():
            raise ValidationError(
                "Company cannot be empty"
            )

    @validates("applied_date")
    def validate_date(self, value, **kwargs):

        if value > date.today():
            raise ValidationError(
                "Date cannot be in future"
            )

File: job-tracker-frontend/src/validation/applicationSchema.js
import { z } from "zod";

export const applicationSchema = z.object({
  company: z
    .string()
    .trim()
    .min(1, "Company is required"),

  role: z
    .string()
    .trim()
    .min(1, "Role is required"),

  status: z
    .string()
    .min(1, "Status is required"),

  appliedDate: z
    .string()
    .min(1, "Applied date is required"),

  notes: z
    .string()
    .max(500, "Notes cannot exceed 500 characters")
    .optional(),
});